### ISSUE TITLE
Out of memory

### ISSUE THREAD

**Issue by jsmouret** [Unity-Technologies/unity-cache-server#114]
Hello,

Our cache server is still running out of memory after a few hours.

Cacher server version: **6.3.0**
Nodejs version: **v10.15.3**
Command line:  `nodejs --max-old-space-size=4096 /usr/lib/node_modules/unity-cache-server/main.js`

Thanks for any help

```
Mar  7 17:18:49 unity3d-cache nodejs[17265]: <--- Last few GCs --->
Mar  7 17:18:49 unity3d-cache nodejs[17265]: [17265:0x397e270]  5318523 ms: Scavenge 4078.3 (4163.3) -> 4077.5 (4163.8) MB, 9.9 / 0.0 ms  (average mu = 0.184, current mu = 0.137) allocation failure
Mar  7 17:18:49 unity3d-cache nodejs[17265]: [17265:0x397e270]  5318537 ms: Scavenge 4078.4 (4163.8) -> 4077.6 (4164.3) MB, 10.1 / 0.0 ms  (average mu = 0.184, current mu = 0.137) allocation failure
Mar  7 17:18:49 unity3d-cache nodejs[17265]: [17265:0x397e270]  5318550 ms: Scavenge 4078.5 (4164.3) -> 4077.8 (4165.3) MB, 10.3 / 0.0 ms  (average mu = 0.184, current mu = 0.137) allocation failure
Mar  7 17:18:49 unity3d-cache nodejs[17265]: <--- JS stacktrace --->
Mar  7 17:18:49 unity3d-cache nodejs[17265]: ==== JS stack trace =========================================
Mar  7 17:18:49 unity3d-cache nodejs[17265]:     0: ExitFrame [pc: 0x13c4ff9dbe1d]
Mar  7 17:18:49 unity3d-cache nodejs[17265]: Security context: 0x00109993c7a9 <JSObject>
Mar  7 17:18:49 unity3d-cache nodejs[17265]:     1: _handleCommand [0x493819da881] [/usr/lib/node_modules/unity-cache-server/lib/server/command_processor.js:~266] [pc=0x13c5001b369f](this=0x029fe723fdf9 <Duplex map = 0xc5c50dcae49>,data=0x05a1d5961f79 <Uint8Array map = 0xe651a1d3771>)
Mar  7 17:18:49 unity3d-cache nodejs[17265]:     2: _write [0x493819da639] [/usr/lib/node_modules/unity-cache-server/lib/server/command_processor.js:~104] [pc=0x13c...
Mar  7 17:18:49 unity3d-cache nodejs[17265]: FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
Mar  7 17:18:49 unity3d-cache nodejs[17265]:  1: 0x8dc510 node::Abort() [/usr/bin/nodejs]
Mar  7 17:18:49 unity3d-cache nodejs[17265]:  2: 0x8dc55c  [/usr/bin/nodejs]
Mar  7 17:18:49 unity3d-cache nodejs[17265]:  3: 0xad9b5e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/usr/bin/nodejs]
Mar  7 17:18:49 unity3d-cache nodejs[17265]:  4: 0xad9d94 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/usr/bin/nodejs]
Mar  7 17:18:49 unity3d-cache nodejs[17265]:  5: 0xec7bf2  [/usr/bin/nodejs]
Mar  7 17:18:49 unity3d-cache nodejs[17265]:  6: 0xec7cf8 v8::internal::Heap::CheckIneffectiveMarkCompact(unsigned long, double) [/usr/bin/nodejs]
Mar  7 17:18:49 unity3d-cache nodejs[17265]:  7: 0xed3dd2 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/usr/bin/nodejs]
Mar  7 17:18:49 unity3d-cache nodejs[17265]:  8: 0xed4704 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/usr/bin/nodejs]
Mar  7 17:18:49 unity3d-cache nodejs[17265]:  9: 0xed7371 v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [/usr/bin/nodejs]
Mar  7 17:18:49 unity3d-cache nodejs[17265]: 10: 0xea07f4 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [/usr/bin/nodejs]
Mar  7 17:18:49 unity3d-cache nodejs[17265]: 11: 0x114018e v8::internal::Runtime_AllocateInNewSpace(int, v8::internal::Object**, v8::internal::Isolate*) [/usr/bin/nodejs]
Mar  7 17:18:49 unity3d-cache nodejs[17265]: 12: 0x13c4ff9dbe1d
```

**Comment by stephen-palmer** [comment]
can you please also post your config using `node main.js --dump-config`?

**Comment by jsmouret** [comment]
Using cache_ram seems to make the problem show up more quickly than cache_fs.
Here is the config;
```
Cache:
  defaultModule: cache_fs
  options:
    processor:
      putWhitelist: []
    cache_ram:
      cachePath: .cache_ram
      pageSize: 200000000
      maxPageCount: 227
      minFreeBlockSize: 1024
      persistence: false
      persistenceOptions:
        autosave: true
        autosaveInterval: 10000
        throttledSaves: false
      highReliability: false
      highReliabilityOptions:
        reliabilityThreshold: 2
        saveUnreliableVersionArtifacts: true
        multiClient: false
    cache_fs:
      cachePath: .cache_fs
      cleanupOptions:
        expireTimeSpan: P30D
        maxCacheSize: 0
      persistence: true
      persistenceOptions:
        autosave: true
        autosaveInterval: 10000
        throttledSaves: false
      highReliability: false
      highReliabilityOptions:
        reliabilityThreshold: 2
        saveUnreliableVersionArtifacts: true
        multiClient: false
Mirror:
  addresses: []
  options:
    queueProcessDelay: 2000
    connectionIdleTimeout: 10000
Server:
  port: 8126
  options:
    allowIpv6: false
Global:
  logLevel: 3
```

**Comment by corngood** [comment]
Hey @stephen-palmer,

As suggested in the other thread, here's a heap snapshot for the process:

```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
13337 unityca+  20   0 3321028 2.149g  11604 S   0.0 13.8   2:58.25 node
```

[Heap-20190327T085422.heapsnapshot.txt](https://github.com/Unity-Technologies/unity-cache-server/files/3012999/Heap-20190327T085422.heapsnapshot.txt)

Maybe you can see some sort of problem in there, but to me it doesn't seem to account for anywhere near 2g of physical memory.

**Comment by stephen-palmer** [comment]
Yeah I don't see evidence of a heap/javascript memory leak. I'm also not yet convinced this is a general/widespread issue based on a relatively small number of problem reports, relative to install base.

I've been searching for similar issues from other node users/apps of native/non-heap memory leaks.  This one might be relevant: https://github.com/nodejs/help/issues/1518

**Comment by timezombi** [comment]
Just to add one to the list, we give Unity Cache Server 49GB of RAM for oldspacesize and we still receive the same errors occasionally. Granted our project size is totally ridiculous.

We just recently switched over to UCS 6.3.0 and node v10.15.3. We haven't experienced a crash yet, but I can report back if we do.

To the original poster: We ended up "solving" the heap space issues by running it in a wrapper script that restarts the unity-cache-server whenever it exits with a non-zero exit code, then it sends me an email alert and starts a new log. It didn't stop the crashes but it did provide a way to rapidly recover when we had problems.

**Comment by jsmouret** [comment]
@timezombi I am using `--workers` to get the "autorestart" feature :)
48GB of RAM and it crashes once a week when a worker gets killed by the oom-killer.
No email notification, just jenkins aborting builds because of timeout.

**Comment by corngood** [comment]
I've been running for a bit with `mtrace`, but as with the v8 snapshot it only accounts for a tiny fraction of the total physical memory in libc allocations.

`pmap` shows a single large physical page range that seems to match up with the unaccounted for memory, but I haven't been able to track down exactly what it is...  It's hard to imagine fragmentation causing this serious a problem, but who knows.

That's all the investigation I have time for this week.  I'll try to get back to it at some point.

**Comment by timezombi** [comment]
Just to follow up on this, we did experience a crash again this morning. Total uptime was 17h 47m 56s. Each time this happens, millions of lines of `Adding file to send queue, size <some_number>` appear in the logs immediately before the GC failure. After it crashed, our recovery/restart script kicked in and everything has been stable since then.

This is the first time this has happened since upgrading to this version of node / Unity Cache server, but we have experienced this many times in older versions.

If there is anything else we can do to try to help you troubleshoot this problem, please let me know!

`...............
[2019-04-11 08:23:48] Adding file to send queue, size 4308
[2019-04-11 08:23:48] Adding file to send queue, size 15837
[2019-04-11 08:23:48] Adding file to send queue, size 2809824
[2019-04-11 08:23:48] Adding file to send queue, size 663
[2019-04-11 08:23:48] Adding file to send queue, size 4276
[2019-04-11 08:23:48] Adding file to send queue, size 18739
[2019-04-11 08:23:48] Adding file to send queue, size 4412
[2019-04-11 08:23:48] Adding file to send queue, size 2824
[2019-04-11 08:24:06]
[2019-04-11 08:24:06] <--- Last few GCs --->
[2019-04-11 08:24:06]
[2019-04-11 08:24:06] [2647:0x36b1210] 64054016 ms: Scavenge 32672.3 (33245.9) -> 32661.7 (33245.9) MB, 403.0 / 0.0 ms  (average mu = 0.915, current mu = 0.949) allocation failure
[2019-04-11 08:24:06] [2647:0x36b1210] 64054512 ms: Scavenge 32672.4 (33245.9) -> 32662.2 (33245.9) MB, 425.0 / 0.0 ms  (average mu = 0.915, current mu = 0.949) allocation failure
[2019-04-11 08:24:06] [2647:0x36b1210] 64055016 ms: Scavenge 32672.4 (33245.9) -> 32662.4 (33245.9) MB, 432.7 / 0.0 ms  (average mu = 0.915, current mu = 0.949) allocation failure
[2019-04-11 08:24:06]
[2019-04-11 08:24:06]
[2019-04-11 08:24:06] <--- JS stacktrace --->`

Cacher server version: 6.3.0
Nodejs version: v10.15.3

Command Line: `node --max-old-space-size=49152 /usr/local/bin/unity-cache-server -l 5 --NODE_CONFIG_DIR=/usr/local/bin/config`

Config Dump:
`Cache:
  defaultModule: cache_fs
  options:
    processor:
      putWhitelist: []
    cache_ram:
      cachePath: .cache_ram
      pageSize: 100000000
      maxPageCount: 10
      minFreeBlockSize: 1024
      persistence: true
      persistenceOptions:
        autosave: true
        autosaveInterval: 10000
        throttledSaves: false
      highReliability: false
      highReliabilityOptions:
        reliabilityThreshold: 2
        saveUnreliableVersionArtifacts: true
        multiClient: false
    cache_fs:
      cachePath: .cache_fs
      cleanupOptions:
        expireTimeSpan: P30D
        maxCacheSize: 0
      persistence: true
      persistenceOptions:
        autosave: true
        autosaveInterval: 10000
        throttledSaves: false
      highReliability: false
      highReliabilityOptions:
        reliabilityThreshold: 2
        saveUnreliableVersionArtifacts: true
        multiClient: false
Mirror:
  addresses: []
  options:
    queueProcessDelay: 2000
    connectionIdleTimeout: 10000
Server:
  port: 8126
  options:
    allowIpv6: false
Global:
  logLevel: 3`

**Comment by stephen-palmer** [comment]
ok, this is the best clue yet. While millions of queued downloads will incur allocations per request, it should only be about 64 bytes per item. I'm actually thinking that debug logging might be the real culprit here. The logging strings are copied around in memory at least a couple of times I'm sure, and though they do eventually garbage collect, it takes some cycles.

Can you take your logging level down one notch and see how it runs for a while?

**Comment by timezombi** [comment]
No problem! I set the logging level to 3 instead of 5. 

With our current oldspacesize, it usually takes several weeks for us to encounter any problems. I'll post here again if/when it happens.

**Comment by stephen-palmer** [comment]
This new branch takes a pass at minimizing string allocs during logging, and especially when the log level is below 5 (debug).

**Comment by timezombi** [comment]
Since setting the log level back down to 3, it has been running pretty solidly for the past 3 weeks. However, we experienced a crash last night, same problem with JavaScript heap out of memory.

> [2019-05-02 18:18:10] Sent 892 of 892 requested files (5295 chunks) totaling 275.55 MB in 16.341 seconds (16.86 MB/sec) to 10.1.0.26:57069
> [2019-05-02 18:19:11] Sent 48 of 48 requested files (398 chunks) totaling 21.91 MB in 6.678 seconds (3.28 MB/sec) to 10.1.1.249:51229
> [2019-05-02 18:21:20] Sent 64 of 64 requested files (211 chunks) totaling 9.14 MB in 3.852 seconds (2.37 MB/sec) to 10.1.0.26:57721
> [2019-05-02 18:33:45] Sent 2 of 2 requested files (2 chunks) totaling 7.41 KB in 0.026 seconds (285.16 KB/sec) to 10.1.1.164:50903
> [2019-05-02 18:34:41] Sent 264 of 264 requested files (1036 chunks) totaling 48.38 MB in 21.113 seconds (2.29 MB/sec) to 10.1.0.31:62299
> [2019-05-02 18:37:39] Sent 2 of 2 requested files (2 chunks) totaling 7.41 KB in 0.004 seconds (1.81 MB/sec) to 10.1.0.31:62441
> [2019-05-02 19:05:23] Sent 4 of 4 requested files (4 chunks) totaling 17.63 KB in 0.007 seconds (2.46 MB/sec) to 10.1.0.34:53508
> [2019-05-02 19:14:33] Sent 2 of 2 requested files (2 chunks) totaling 9.03 KB in 0.006 seconds (1.47 MB/sec) to 10.1.0.34:53530
> [2019-05-02 19:22:55] Sent 10 of 10 requested files (27 chunks) totaling 1.08 MB in 0.384 seconds (2.82 MB/sec) to 10.1.1.126:65470
> [2019-05-02 19:34:05] Sent 2 of 2 requested files (2 chunks) totaling 8.6 KB in 0.029 seconds (296.5 KB/sec) to 10.1.0.34:53634
> [2019-05-02 19:37:27] Sent 2 of 2 requested files (2 chunks) totaling 8.6 KB in 0.004 seconds (2.1 MB/sec) to 10.1.0.34:53682
> [2019-05-02 19:48:39] Sent 14 of 14 requested files (49 chunks) totaling 2.14 MB in 0.94 seconds (2.28 MB/sec) to 10.1.1.126:65493
> [2019-05-02 19:55:36] Sent 88 of 88 requested files (524 chunks) totaling 27.07 MB in 10.305 seconds (2.63 MB/sec) to 10.1.0.34:53850
> [2019-05-02 19:56:46] Sent 54 of 54 requested files (1413 chunks) totaling 84.71 MB in 2.933 seconds (28.88 MB/sec) to 10.1.0.34:53874
> [2019-05-02 20:10:37] Sent 2 of 2 requested files (2 chunks) totaling 4.93 KB in 0.152 seconds (32.42 KB/sec) to 10.1.1.249:52551
> [2019-05-02 20:30:27]
> [2019-05-02 20:30:27] <--- Last few GCs --->
> [2019-05-02 20:30:27]
> [2019-05-02 20:30:27] [7739:0x26d2210] 359384827 ms: Scavenge 31501.5 (33033.0) -> 31486.7 (33033.0) MB, 119.0 / 0.0 ms  (average mu = 0.953, current mu = 0.953) allocation failure
> [2019-05-02 20:30:27] [7739:0x26d2210] 359385064 ms: Scavenge 31501.9 (33033.0) -> 31487.2 (33033.0) MB, 141.3 / 0.0 ms  (average mu = 0.953, current mu = 0.953) allocation failure
> [2019-05-02 20:30:27] [7739:0x26d2210] 359385316 ms: Scavenge 31502.1 (33033.0) -> 31487.8 (33033.0) MB, 159.9 / 0.0 ms  (average mu = 0.953, current mu = 0.953) allocation failure
> [2019-05-02 20:30:27]
> [2019-05-02 20:30:27]
> [2019-05-02 20:30:27] <--- JS stacktrace --->
> [2019-05-02 20:30:27] Cannot get stack trace in GC.
> [2019-05-02 20:30:27] FATAL ERROR: NewSpace::Rebalance Allocation failed - JavaScript heap out of memory
> [2019-05-02 20:30:28]  1: 0x8dc510 node::Abort() [node]
> [2019-05-02 20:30:28]  2: 0x8dc55c  [node]
> [2019-05-02 20:30:28]  3: 0xad9b5e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
> [2019-05-02 20:30:28]  4: 0xad9d94 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
> [2019-05-02 20:30:28]  5: 0xec7bf2  [node]
> [2019-05-02 20:30:28]  6: 0xefaf66 v8::internal::MarkCompactCollector::Evacuate() [node]
> [2019-05-02 20:30:28]  7: 0xefb5c2 v8::internal::MarkCompactCollector::CollectGarbage() [node]
> [2019-05-02 20:30:28]  8: 0xed3411 v8::internal::Heap::MarkCompact() [node]
> [2019-05-02 20:30:28]  9: 0xed3b01 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [node]
> [2019-05-02 20:30:28] 10: 0xed4704 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
> [2019-05-02 20:30:28] 11: 0xed7371 v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [node]
> [2019-05-02 20:30:28] 12: 0xea07f4 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [node]
> [2019-05-02 20:30:28] 13: 0x11402db v8::internal::Runtime_AllocateInTargetSpace(int, v8::internal::Object**, v8::internal::Isolate*) [node]
> [2019-05-02 20:30:28] 14: 0x1d244dbe1d

**Comment by stephen-palmer** [comment]
For the next release, I made a change to how the download queue is managed and processed, which should result in a lower memory high-watermark. The worst case scenario is when a client wants to download an entire project from the cache, queuing up 10s or 100s of thousands of download requests.

In those cases, especially in the 100s of thousands, performance and memory are both significantly impacted, so this change addresses both.

The change is in this PR: https://github.com/Unity-Technologies/unity-cache-server/pull/120

**Comment by stephen-palmer** [comment]
Addressed in v6.4.0. Beta release is available: https://github.com/Unity-Technologies/unity-cache-server/releases/tag/v6.4.0-beta0

**Comment by timezombi** [comment]
Excellent! I will try deploying it this weekend. Thank you for all your hard work on these fixes!

**Comment by timezombi** [comment]
I tried running 6.4.0b2 with log level 4 in a test environment, allocating 6GB for oldspacesize (production is 49GB). The platforms I used for testing were PS4 and PC. Running the special asset uploader script you sent us a while back seeded the cache with about 100,000 assets per platform just fine, for a total of ~200,000 assets.

When I tried to start a full import of a project, the familiar debug logging messages of "Adding file to send queue" started spamming the screen. After 129080 lines of that, the cache server hangs (without error) and the client eventually pops up a message saying it is unable to connect to cache server. I'm not sure if this is an actual Out Of Memory error or something else. 

After the server hang, I killed the cache server and launched it with log level 3, where the project finished its import after hitting the reconnect button.

If I had a choice, I would prefer the cache server just crash instead of hang in this scenario since it is easier to recover from and doesn't obviously stall all the clients from importing. I only let it hang for 10-15 minutes, so I'm not sure if it would have eventually recovered. For now, I no longer intend on testing this in log level 4 except for very small, specific cache server requests.

In today's round of testing (part 2), I started some full imports of our project on a single connected client (Unity 2018.3.11f1) and the server set to log level 3 instead. If all goes well, by tomorrow morning there should have been 3 clients that succeeded in full imports. I will also try to test multiple simultaneous full imports to see how things go.

If the final version of 6.4.0 isn't out by Memorial Day, I may have an opportunity to stress test the beta with multiple simultaneous full imports with production level resources (20gbps fiber channel network, 49GB oldspacesize, 6 logical cores, etc) sometime that weekend. These are the sorts of workloads where I've seen OOM crashes with debug logging turned off.

In summary, aside from the server hang, the cache server has not been very problematic. The memory usage and stability has actually been quite good. If the server hang repros with debug logging turned off, we will have some serious reservations about deploying this to production but are otherwise happy with the small testing we have done.

I will report back again after our current round of testing completes.

**Comment by timezombi** [comment]
I did some major stress testing on the beta cache server this week. The results of the testing has given me enough confidence to deploy 6.4.0b3 into production this weekend if a newer beta or final version is not released. I was not able to repro _any_ OOM errors despite simulating workloads far greater than what we could ever realistically produce in production.

It seems like 6.4.0 scales much better than 6.3.0 when handling simultaneous large transactions. With ludicrously large simulated workloads, memory usage did not go over 2.2GB total. On 6.3.0, we would run out of memory (48GB) with comparatively modest workloads.

Memory usage in general was also at least 50% better than 6.3.0 when idling.

Frankly, I feel like we are 99% out of the woods now. I'll run it with `--diag-client-recorder` enabled for the next couple of weeks just in case.

I'm writing up the full extent of my testing and will post it in Issue #135 shortly.

**Comment by timezombi** [comment]
I was able to catch Unity Cache Server 6.3.0 in the act, right before running out of memory. I had looked at htop this morning and didn't notice anything unusual about the memory usage. When I looked again about an hour later, it was using up all of the physical RAM and got through almost all of the available swap space (for a total of 85.3GB). This was a surprise since it was more than what I was assigning for max-old-space-size. This is when I realized that Unity Cache Server may be setting max-old-space-size for each thread running. Since we had 25GB assigned to max-old-space-size and there were about 10 or 11 threads running concurrently, I may have unintentionally over-allocated the available memory on my machine. However, I don't know if setting this to a smaller value would have actually prevented an Out-Of-Memory error. I will reduce the max-old-space-size this weekend.

Immediately prior to looking at the resource usage, someone had just triggered the null-ref bug (as seen in #100 and #95) which causes their local Editor.log to infinitely display "unexpected response from cache server". One of our engineers discovered some client code mishandles some exceptions which exacerbates the problem. He filed a ticket with Unity about it here: https://support.unity3d.com/hc/en-us/requests/665054 

The misbehaving Unity client in question had been stuck in an infinite loop for about 15 or 20 minutes. I killed his Editor sometime around 11:05-11:15 then switched him over to the beta cache server to avoid encountering it again. 

When I went back to my workstation, I had noticed the extreme memory usage. The server went on to run normally (with 85GB of RAM allocated) for several more hours before I killed it.

This whole episode occurred when there was very little activity from the cache server, no crazy large requests to flood the download queue or anything. The last hours of transactions before the heavy memory usage looked like a typical workload but with some Errors thrown in that were probaby caused by killing the Unity Editor.

Given the circumstances surrounding the high memory event, I suspect that the OOM errors I have been experiencing were caused by the null-ref issue (#100) and that the issue will be fixed when switching to 6.4.0. We are switching to 6.4.0 this weekend. I will follow up again if still experiencing OOM issues.

For reference, our current cache has ~292k assets stored with about ~309GB total disk space being used. 

Attached is the log for several hours before and after the high memory usage:
[unity_high_memory.log](https://github.com/Unity-Technologies/unity-cache-server/files/3315981/unity_high_memory.log)

This probably doesn't help much but here is the pmap:
[unity_high_mem_pmap.log](https://github.com/Unity-Technologies/unity-cache-server/files/3316007/unity_high_mem_pmap.log)

I didn't have anything more sophisticated such as `mtrace` set up, but I hope this circumstantial evidence helps somewhat.

**Comment by stephen-palmer** [comment]
High memory issues are resolved to the best of my knowledge in v6.4.0 (beta).
